### PR TITLE
feat(wasinn): add support for RWKV backend

### PR DESCRIPTION
**Pull request by PhantomInTheWire**
relevant issue: https://github.com/WasmEdge/WasmEdge/issues/3124
just a heads up some things are different from other backends
- RWKV utilizes an external hugging face tokenizer via tokenizers-cpp.
- rwkv.cpp provides raw logits only, so we implement our own sampling pipeline inspired by llama.cpp's approach.
- RWKV contexts share the model weights but have independent state buffers. Cloning allows multiple inference sessions with isolated states while sharing the model memory.

### CODE REVIEW

**Comment by PhantomInTheWire** [comment]
@hydai @dm4 @ibmibmibm can you please review this one? thanks.

**Comment by hydai** [comment]
> @hydai @dm4 @ibmibmibm can you please review this one? thanks.

Do you have a test case or example to demonstrate that this is workable?  
Have you added at least a minimum test case to the tests folder?  
If not, there is nothing to review.

**Comment by PhantomInTheWire** [comment]
@hydai I do have an example that works but running it requires quite a bit of setup(didn't add it initially because of that and because no other backend had a dedicated example), I'll push a Makefile that simplifies things for the example in a second.
also im not sure if you missed it but I did have tests in the test folder

https://github.com/user-attachments/assets/c8c0fb53-904d-47db-8970-6f71e1962541

**Comment by PhantomInTheWire** [comment]
@hydai I have pushed the example, it can be run by 
```bash
# make sure you have wasi sdk and the branch built with the rwkv backend
cd examples/wasm
make run PROMPT="hello rwkv"
```

I am not really sure what the best place to put the example was though.

**Comment by PhantomInTheWire** [comment]
@hydai did you have a chance to look at this yet?

**Comment by hydai** [comment]
I prefer to add the examples at https://github.com/second-state/WasmEdge-WASINN-examples/, aligned with the other examples.

File: cmake/WASINNDeps.cmake [hunk 1]
```diff
@@ -700,3 +706,89 @@ function(wasmedge_setup_bitnet_target target)
...
+    else()
+      message(STATUS "WASI-NN RWKV backend: Disable RWKV_CUBLAS")
+      set(RWKV_CUBLAS OFF CACHE BOOL "Enable cuBLAS for RWKV")
+    endif()
+
+    message(STATUS "Downloading rwkv.cpp source")
+    FetchContent_Declare(
+      rwkv
+      GIT_REPOSITORY https://github.com/RWKV/rwkv.cpp.git
+      GIT_TAG        master-14663c8  # Pinned for reproducibility (Mar 23, 2025)
```
Comment (hydai) [inline cmake/WASINNDeps.cmake:732]: Just want to check if there is a tag or release on RWKV.

File: plugins/wasi_nn/wasinntypes.h [hunk 2]
```diff
@@ -52,6 +52,7 @@ enum class Backend : uint8_t {
   ChatTTS = 12,
   OpenVINOGenAI = 13,
   BitNet = 14,
+  RWKV = 15,
```
Comment (hydai) [inline plugins/wasi_nn/wasinntypes.h:55]: You should leave comments here to indicate the index: https://github.com/WasmEdge/WasmEdge/issues/3491.

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+#include <cstdint>
+#include <ios>
+#include <string>
+#include <string_view>
+#include <vector>
+
+#ifdef WASMEDGE_PLUGIN_WASI_NN_BACKEND_RWKV
+#include "simdjson.h"
+
+#include <algorithm>
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: Duplicated.

**Comment by PhantomInTheWire** [comment]
sure I'll open a follow up pr to https://github.com/second-state/WasmEdge-WASINN-examples once this is merged(and remove the example once I get some time)

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+        std::filesystem::path(ModelPath).parent_path();
+    if (ModelDir.empty()) {
+      ModelDir = ".";
+    }
+
+    const std::vector<std::string> TokenizerCandidates = {
+        "tokenizer.json", "20B_tokenizer.json", "rwkv_vocab_v20230424.json",
+        "vocab.json"};
+
+    const bool TokenizerLoaded = [&]() {
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: I don't understand the purpose of this lambda function; you define it and use it immediately, and moreover, it is used only once. Why is this necessary?

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 5]
```diff
@@ -0,0 +1,861 @@
...
+  }
+  return {};
+}
+
+} // namespace
+
+Expect<ErrNo> load(WasiNNEnvironment &Env, Span<const Span<uint8_t>> Builders,
+                   [[maybe_unused]] Device Device, uint32_t &GraphId) noexcept {
+  if (Builders.size() < 1) {
+    spdlog::error("[RWKV] Requires at least 1 builder (model path)"sv);
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: You used the `LOG_` series functions everywhere else; why are you using `spdlog::error` here?

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 6]
```diff
@@ -0,0 +1,861 @@
...
+                         Builders[1].size());
+    auto Res = parseMetadata(Graph.RWKVConfig, Metadata);
+    if (Res != ErrNo::Success) {
+      Env.NNGraph.pop_back();
+      return Res;
+    }
+    if (Graph.RWKVConfig.MainGPU != 0) {
+      LOG_WARN("RWKV backend currently only supports a single GPU. 'main-gpu' "
+               "parameter will be ignored."sv);
+    }
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: If this parameter will be ignored, why do we need it in the parser?

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 7]
```diff
@@ -0,0 +1,861 @@
...
+            }
+          }
+        }
+      }
+      return false;
+    }();
+
+    if (!TokenizerLoaded) {
+      LOG_ERROR(
+          "No tokenizer found. Please provide tokenizer path "
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: What's the reason we need to add the auto-discovered mechanism for the tokenizer? Why can't users send it to us?

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 8]
```diff
@@ -0,0 +1,861 @@
...
+  GraphId = static_cast<uint32_t>(Env.NNGraph.size() - 1);
+  Env.NNGraph.back().setReady();
+
+  return ErrNo::Success;
+}
+
+Expect<ErrNo> initExecCtx(WasiNNEnvironment &Env, uint32_t GraphId,
+                          uint32_t &ContextId) noexcept {
+  if (GraphId >= Env.NNGraph.size()) {
+    spdlog::error("[RWKV] Invalid graph ID: {}"sv, GraphId);
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: Ditto.

**Review by hydai — CHANGES_REQUESTED** [review]

File: cmake/WASINNDeps.cmake (hunk 1 above)
Comment (PhantomInTheWire) [reply cmake/WASINNDeps.cmake:732]: image
this seems to be the latest release idk why they would call their latest release master-hash but it is what it is

File: plugins/wasi_nn/wasinntypes.h (hunk 2 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinntypes.h:55]: sure, done.

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 3 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: removed.

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 5 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: my bad. refactored.

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 6 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: sure we can git rid of it.

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 7 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: makes sense. fixed

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 8 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: my bad. fixed

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 4 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: removed it

**Comment by PhantomInTheWire** [comment]
@hydai When you have a moment, PTAL

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 9]
```diff
@@ -0,0 +1,803 @@
...
+  if (Info) {                                                                  \
+    spdlog::info("[WASI-NN] RWKV backend: "sv __VA_ARGS__);                    \
+  }
+
+#define LOG_WARN(...) spdlog::warn("[WASI-NN] RWKV backend: "sv __VA_ARGS__);
+
+#define LOG_ERROR(...) spdlog::error("[WASI-NN] RWKV backend: "sv __VA_ARGS__);
+
+// Implementation inspired by llama.cpp sampling logic:
+// https://github.com/ggml-org/llama.cpp
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp]: Can you indicate the exact location of the sampling logic you are referring to? It should include the commit hash and path so I can verify whether the implementation is correct.

File: plugins/wasi_nn/wasinn_rwkv.h [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+  bool EnableLog = false;
+  bool EnableDebugLog = false;
+
+  uint64_t MaxTokens = 256;
+  float Temperature = 1.0f;
+  uint64_t TopK = 40;
+  float TopP = 0.5f;
+  uint64_t MinKeep = 1;
+  float PresencePenalty = 0.0f;
+  float FrequencyPenalty = 0.0f;
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.h]: You have some default values here. Could you show the reference indicating that RWKV makes the same assumption?

**Review by hydai — CHANGES_REQUESTED** [review]

File: plugins/wasi_nn/wasinn_rwkv.cpp (hunk 9 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.cpp]: https://github.com/ggml-org/llama.cpp/blob/b7feacf7f39d79518e6ca48a9fcf697c5244b585/common/sampling.cpp, I've added the same in comments

File: plugins/wasi_nn/wasinn_rwkv.h (hunk 10 above)
Comment (PhantomInTheWire) [reply plugins/wasi_nn/wasinn_rwkv.h]: I just picked sensible defaults based on the llama.cpp implementation, I couldn't exactly find recommendations from rwkv.cpp, but there is a example from them https://github.com/RWKV/rwkv.cpp/blob/master/python/chat_with_bot.py it has slightly different PresencePenalty FrequencyPenalty and Temperature from defaults I picked, so I changed mine to match there example

**Comment by PhantomInTheWire** [comment]
ping @hydai in case this got buried

**Comment by PhantomInTheWire** [comment]
@hydai did you have a chance to look at this yet?

**Comment by hydai** [comment]
Hi @PhantomInTheWire,  
I am too busy to review this right now, as I need to check the llama.cpp sampling method first and then review your implementation. This PR will be pending due to the capacity. Once I have enough time, I will go back to it.

**Comment by PhantomInTheWire** [comment]
@hydai sure that's fine, I have noticed the sudden influx of prs too, I'll help out with the first pass of reviews where I can.

**Comment by hydai** [comment]
Hi @PhantomInTheWire,  
Just wanted to let you know that most of our maintainers will be out of the office due to the Lunar New Year vacation. Since this PR is too large to be reviewed before the end of this weekday, I plan to process it after I return to the office. Thanks for your patience.

**Comment by PhantomInTheWire** [comment]
@hydai Wishing you all a happy Lunar New Year. Enjoy the holiday, and I’ll look forward to your review once you’re back.

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 11]
```diff
@@ -0,0 +1,803 @@
...
+  uint32_t GraphId = CxtRef.GraphId;
+
+  if (GraphId >= Env.NNGraph.size()) {
+    LOG_ERROR("computeSingle: Invalid graph ID in context: {}"sv, GraphId);
+    return ErrNo::InvalidArgument;
+  }
+
+  auto &Graph = Env.NNGraph[GraphId].get<Backend::RWKV>();
+
+  if (!Graph.RWKVCtx) {
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp:608]: I found you used `RWKVCtx` in both `Graph` and `Context`, what's the difference?
It's really confused for people.

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 12]
```diff
@@ -0,0 +1,803 @@
...
+
+  rwkv_init_state(CxtRef.RWKVCtx.get(), CxtRef.StateBuffer.data());
+
+  Env.NNContext[ContextId].setReady();
+
+  return ErrNo::Success;
+}
+
+Expect<ErrNo> setInput(WasiNNEnvironment &Env, uint32_t ContextId,
+                       uint32_t Index [[maybe_unused]],
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp:405]: You mark `Index` as `maybe_unused` but it's definitely used in line 416.

File: plugins/wasi_nn/wasinn_rwkv.cpp [hunk 13]
```diff
@@ -0,0 +1,803 @@
...
+    }
+    LOG_INFO(Graph.RWKVConfig.EnableLog, "Loaded tokenizer from: {}"sv,
+             Graph.RWKVConfig.TokenizerPath);
+  } else {
+    std::filesystem::path ModelDir =
+        std::filesystem::path(ModelPath).parent_path();
+    if (ModelDir.empty()) {
+      ModelDir = ".";
+    }
+
```
Comment (hydai) [inline plugins/wasi_nn/wasinn_rwkv.cpp:350]: This section seems to be a dead code. You defined `ModelDir` without using it.

**Review by hydai — CHANGES_REQUESTED** [review]

**Comment by hydai** [comment]
Thanks for the contribution. Please feel free to re-open it once you continue working on it.
